### PR TITLE
Bug 1991573: [NETOBSERV-29] Turn on strictNullCheck on network-policies files

### PR DESCRIPTION
**Pull request by jotak**
New tsconfig in `frontend/packages/console-app/src/components/network-policies` defines stricter rules, essentially for strict-null-checks (increases type safety). Also uses `noImplicitAny`.

JIRA https://issues.redhat.com/browse/NETOBSERV-29

Signed-off-by: Joel Takvorian <jtakvori@redhat.com>

### CODE REVIEW

**Comment by jotak** [comment]
cc @mariomac

**Review by mariomac — APPROVED** [review]

**Comment by mariomac** [comment]
LGTM

File: frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx
Comment (andrewballantyne) [inline frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx:7]: Small nit, doesn't mean a lot today but is future balanced if we go with the props of `AsyncComponent`. You can omit the static prop that is provided:

```suggestion
const NameValueEditorComponent = (props: Omit<AsyncComponentProps, 'loader'>) => (
```

File: frontend/packages/console-app/src/components/network-policies/network-policy-model.ts
Comment (andrewballantyne) [inline frontend/packages/console-app/src/components/network-policies/network-policy-model.ts:145]: Hmmm, did TypeScript recommend more typing? Because I would think you are good with the previous design. Which is a tad bit more future proof and I'm not sure you're adding much here as `NetworkPolicyKind` already types `policyTypes`.

File: frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx
Comment (andrewballantyne) [inline frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx:25]: Just a piece of information, this isn't related to your changes so I'm not necessarily asking for this change.

We should avoid passing `t` functions into methods. Any method can invoke the static `i18next.t` function safely as the configuration shared between the `t` functions is the same. It would also clean up method signatures from having `t` unnecessarily required.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx
Comment (kdoberst) [reply frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx:7]: I agree. It is preferable to avoid type of `any`.

Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx:7]: I tried to fix that `any` before, but the really interesting type here is the one of `NameValueEditor` (which is JSX+proptypes, unfortunately it hasn't been typescriptified - I briefly tried to do that, but gave up as it was turning into a huge refactoring).

If you look at the `AsyncComponentProps`, it's defined as:
```
{
  loader: () => Promise<React.ComponentType>;
  LoadingComponent?: React.ReactNode;
} & any;
```
which is basically equal to `any` if we omit `loader`. So, yes I agree to do that change, unfortunately it's not particularly useful :(

File: frontend/packages/console-app/src/components/network-policies/network-policy-model.ts
Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-model.ts:145]: Yes I had to do that to fix a compilation error. Problem was line 157 where `res.spec.policyTypes.push(networkPolicyTypeIngress);` triggered `res.spec.policyTypes possibly undefined` (despite being explicitly defined above but it seems like typescript doesn't fully check for non-nullity of nested types?)

My first intent was to write `res.spec.policyTypes!.push(networkPolicyTypeIngress);` which passes the TS compilation however it triggers a linting error later (`!` is apparently forbidden), so I preferred to rewrite rather than ignoring the linting error.

File: frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx
Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx:25]: Ok I'll rewrite

Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx:25]: Hmm, can you elaborate on invoking the static `i18next.t` ? Do you mean by calling `const { t } = useTranslation();` ?

If so, this change is not as straightforward as it seems. When I do so, I get a react warning:

```
Warning: React has detected a change in the order of Hooks called by NetworkPolicyRuleConfigPanel. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks

   Previous render            Next render
   ------------------------------------------------------
1. useContext                 useContext
2. useState                   useState
3. useRef                     useRef
4. useEffect                  useEffect
5. undefined                  useContext
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

NetworkPolicyRuleConfigPanel@http://localhost:9000/static/create-network-policy-3563d8cb2c6780af5540.js:1752:75
```

It's the first time I see this one, I understand that calls like `useContext` (which is called within `useTranslation`) should not be done conditionally, which is the case here? I guess then I'd need to refactor the component to split in more pieces so that `useTranslation` is called only one time per component.

Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx:25]: ok, forget what I said, I didn't know we could invoke directly `i18next.t` :-)
Can you explain when it's preferable to use `useTranslation` then?

**Comment by jotak** [comment]
@andrewballantyne thanks for the comments, modifications done

File: frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx
Comment (andrewballantyne) [reply frontend/packages/console-app/src/components/network-policies/network-policy-conditional-selector.tsx]: I agree it's not particularly useful today. But in an app this large, types can change and we may turn that into something else over time. In which case your component will still be typed as `any` despite having extra requirements. It's best we lean into types as we have them. Let the compiler help us do the checking when types change.

File: frontend/packages/console-app/src/components/network-policies/network-policy-model.ts
Comment (andrewballantyne) [reply frontend/packages/console-app/src/components/network-policies/network-policy-model.ts:145]: Ah thanks for the information. There isn't anything innately wrong with what you did here... my only concern was if later this `const policyTypes` changes to `let policyTypes` you may get yourself into the loss of an object reference and break the updating logic.

It's probably highly unlikely this will happen so I suppose we can go with this change. This file is somewhat concerning with this reliance on object mutation, but that's a different topic for a different PR/issue.

File: frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx
Comment (andrewballantyne) [reply frontend/packages/console-app/src/components/network-policies/network-policy-rule-config.tsx]: Ah apologizes for the lack of clarity in my original comment. 

- `useTranslation` is a hook based one and thus should be used inside components
- `i18next.t` invoked directly should only be done in utilities (functions that are not components); it pulls from the same config and thus when invoked at runtime will have the correct language setting from the other parts of the UI

A side note, don't use `i18next.t` in the root of any file. The files get parsed once at the first import and thus will lock in the language setting and never change.

**Review by andrewballantyne — APPROVED** [review]

**Comment by andrewballantyne** [comment]
> LGTM

Was already given a LGTM by someone else, so I'll just add the label now that I've approved the PR.

/lgtm

**Comment by andrewballantyne** [comment]
@jotak We require a Bugzilla at this point in the release cycle. Could you log one and link it to the PR? If you have not done this before I'd be happy to give you more information.

File: frontend/packages/console-app/src/components/network-policies/network-policy-model.ts
Comment (jotak) [reply frontend/packages/console-app/src/components/network-policies/network-policy-model.ts:145]: To be honest I'd be in favour of removing the linter rule that forbids `!`, at some point; although the intent is good, but the TS flow control being not perfect, in practice it's common to see cases where it's counter-productive, like here or in some constructs with callbacks.

**Comment by jotak** [comment]
@andrewballantyne bugzilla created: https://bugzilla.redhat.com/show_bug.cgi?id=1991573

**Comment by andrewballantyne** [comment]
> @andrewballantyne bugzilla created: https://bugzilla.redhat.com/show_bug.cgi?id=1991573

You'll need to set the fix version to `4.9.0` to be valid for the master branch. And you need to add the `:` after the number.

**Comment by andrewballantyne** [comment]
/hold

Waiting on 4.10 to open up.

**Comment by jotak** [comment]
I guess it can be merged now? (it was on hold until 4.10 opens up) cc @andrewballantyne

**Comment by andrewballantyne** [comment]
> I guess it can be merged now? (it was on hold until 4.10 opens up) cc @andrewballantyne

Yup looks good, `master` is now 4.10. 

/unhold

**Comment by andrewballantyne** [comment]
Oh right, so the process changed between 4.9 and 4.10 @jotak 

Every issue needs a Bugzilla now or is considered a feature and needs approvals from QE, Docs and the PX teams. Apologizes for the inconvenience. We could probably just bring back your Bugzilla ticket and use that I think

**Comment by andrewballantyne** [comment]
/retitle Bug 1991573: [NETOBSERV-29] Turn on strictNullCheck on network-policies files

**Pull request merged**
